internal/server/auth/middleware/grpc/middleware.go

```diff
@@ -91,10 +113,114 @@ type SkipsAuthenticationServer interface {
 	SkipsAuthentication(ctx context.Context) bool
 }
 
-// UnaryInterceptor is a grpc.UnaryServerInterceptor which extracts a clientToken found
+// AuthenticationRequiredInterceptor is a grpc.UnaryServerInterceptor which requires that
+// all requests contain an Authentication instance on the context.
+func AuthenticationRequiredInterceptor(logger *zap.Logger, o ...containers.Option[InterceptorOptions]) grpc.UnaryServerInterceptor {
+	var opts InterceptorOptions
+	containers.ApplyAll(&opts, o...)
+
+	return func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
+		// skip auth for any preconfigured servers
+		if skipped(ctx, info.Server, opts) {
+			logger.Debug("skipping authentication for server", zap.String("method", info.FullMethod))
+			return handler(ctx, req)
+		}
+
+		auth := GetAuthenticationFrom(ctx)
+		if auth == nil {
+			logger.Error("unauthenticated", zap.String("reason", "authentication required"))
+			return ctx, ErrUnauthenticated
+		}
+
+		return handler(ctx, req)
+	}
+}
+
+// JWTInterceptorSelector is a grpc.UnaryServerInterceptor which selects requests
+// which contain a JWT in the authorization header.
+func JWTInterceptorSelector() selector.Matcher {
+	return selector.MatchFunc(func(ctx context.Context, _ interceptors.CallMeta) bool {
+		md, ok := metadata.FromIncomingContext(ctx)
+		if !ok {
+			return false
+		}
+
+		_, err := jwtFromMetadata(md)
+		return err == nil
+	})
+}
+
+func JWTAuthenticationInterceptor(logger *zap.Logger, validator jwt.Validator, expected jwt.Expected, o ...containers.Option[InterceptorOptions]) grpc.UnaryServerInterceptor {
+	var opts InterceptorOptions
+	containers.ApplyAll(&opts, o...)
+
+	return func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
+		// skip auth for any preconfigured servers
+		if skipped(ctx, info.Server, opts) {
+			logger.Debug("skipping authentication for server", zap.String("method", info.FullMethod))
+			return handler(ctx, req)
+		}
+
+		md, ok := metadata.FromIncomingContext(ctx)
+		if !ok {
+			logger.Error("unauthenticated", zap.String("reason", "metadata not found on context"))
+			return ctx, ErrUnauthenticated
+		}
+
+		token, err := jwtFromMetadata(md)
+		if err != nil {
+			logger.Error("unauthenticated",
+				zap.String("reason", "no authorization provided"),
+				zap.Error(err))
+
+			return ctx, ErrUnauthenticated
+		}
+
+		// TODO: map claims to auth metadata?
+		_, err = validator.ValidateAllowMissingIatNbfExp(ctx, token, expected)
```

Comment:
Should we (instead) always require an expiration by default?

Avoid folks creating infinite tokens. Or optionally, making supporting that an explicit opt-in (InfiniteTokens: true with a warning).

Reply:
Yeah I think we could: https://github.com/hashicorp/cap/blob/1d58e0e8bd5f948b9d2bebe8b25a81eb0288fe32/jwt/jwt.go#L163-L165

I think we can require that at least `exp` is set/valid